### PR TITLE
overview of project components

### LINKED ISSUE #111: Create a small overview of project components

### PR DESCRIPTION
**Pull request by DonatJR**
fixes #111
* small typo and code placement fix (has nothing to do with the issue)
* remove old and outdated documents from previous team project
* add project overview document and mention it in the readme
* show sequence diagram of typical invocation of shark cage in overview

### CODE REVIEW

File: project_overview.md [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+
+* __SharkCageInstallerHelper__: Loaded by the installer. Contains functions which further customize the installer behaviour with functionality not available by default.
+
+## SharkCageInstaller
+
+Installer project which is able to configure target systems for usage of Shark Cage (either user guided or silently).
+
+# Sequence diagram of a typical invocation of Shark Cage
+
+![](sharkcage_sequence_diagramm.png)
```
Comment (SailReal) [inline project_overview.md]: Maybe we can replace this link with that one: `sharkcage_sequence_diagramm`

Then we can delete the image from the repo: https://github.com/SharkCagey/HTWG_shark_cage/blob/111-component-overview/sharkcage_sequence_diagramm.png
I think we should keep as few non-source files as possible directly in the repo.

File: project_overview.md [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+* __CageConfiguratorForm__: Lets the user configure applications to be run inside Shark Cage. The following options are available:
+  * Application (mandatory): Path to the application to be started inside the Shark Cage.
+  * Option to restrict returning from secure desktop to the exit button on the `CageLabeler` (if not checked the desktop will also close once all applications on it are closed). This option exists because there can be compatibility issues with some programs (e.g. Internet Explorer) where checking for the running program is not possible and the secure desktop would close immediately after opening.
+  * Secondary application: Sometimes it is useful to have an auxiliary program running besides the main application. This option currently only supports `Keepass`.
```
Comment (SailReal) [inline project_overview.md]: I would remove the ~only~ to "This option currently supports `Keepass`."

File: project_overview.md [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+  * Picture token (mandatory) for identifying the Shark Cage instance when run (picture is also displayed inside the secure desktop by the `CageLabeler`). This can either be from the users storage or directly captured by a connectec webcam.
```
Comment (SailReal) [inline project_overview.md]: typo, should be connected

File: project_overview.md [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+* __CageLabeler__: Displays information about the currently running configuration: picture token, application names and hashes, activation buttons for the applications (to bring them in focus or restart them), exit restriction information, exit button.
+* __CageManager__: Immediately after getting started the `CageManager` listens for a tcp message. If no message arrives it just exits without doing anything else. If there is an expected message it parses it and then validates (signature / hash) all binaries specified in the reveived configuration. After validation it creates the security attributes with the `SecuritySetup`, creates a new access token with the `tokenLib` and a new desktop with a unique name & the security attributes. After this, the `CageLabeler` is used to display information on this new desktop. Once all this preparation is done the actual processes get started on the secure desktop and the `CageManager` waits for the correct exit conditions before cleaning up and exiting itself.
```
Comment (SailReal) [inline project_overview.md]: ...it creates the security attributes... should be `SECURITY_ATTRIBUTES` or?

Comment (SailReal) [inline project_overview.md]: should be: ...and a new desktop with a unique name ~&~ and the security...

File: project_overview.md [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+* __FullWorkArea__: Used by the `CageDesktop` class to initialize and reserve space for the `CageLabeler`.`
```
Comment (SailReal) [inline project_overview.md]: At the end of the line, a ` is too much

File: project_overview.md [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+* __SecuritySetup__: Sets up and populates a [SECURITY_ATTRIBUTES](https://msdn.microsoft.com/en-us/library/windows/desktop/aa379560%28v=vs.85%29.aspx?f=255&MSPPError=-2147217396) struct used for example to create the secure desktop.
+
+## CageService
+
+* __CageService__: Checks if configuration files gotten as messages from `CageChooser` (and passed on by `CageServiceMain`) have the right access rights, asserts if the `CageManager` binary has the correct certificate and then starts it in the context of the current user (passing on the configuration file as well). Afterwards it waits until the `CageManager` instance closes again and only then accepts further messages (ensuring only one secure desktop can run at the same time).
+* __CageServiceMain__: Starts the `CageService`, continuously waits for messages to arrive (& pass them to the `CageService` instance) and waits for shutdown event in case the windows service gets stopped.
```
Comment (SailReal) [inline project_overview.md]: Same as above: ...arrive (~&~ and pass them...

File: project_overview.md [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+* __NetworkManager__: Handles network communication between the different binaries. When creating an instance a port used for listening must be specified, but outgoing messages can be addressed to all ports used by any of the other binaries / projects. After sending a message the `NetworkManager` briefly waits for a response and reports the result to the caller.
+
+### TokenLib
+
+* __groupManipulation__: Used for creating and deleting local user groups.
+* __tokenManipulation__: Used for manipulating windows access tokens.
+
+### Misc
+
+* __CageData__: Contains a struct which encapsulates all data used by various functions responsible for displaying data / creating processes in the secure desktop.
```
Comment (SailReal) [inline project_overview.md]: should be ... "on the secure desktop." or?

File: project_overview.md (hunk 4 above)
Comment (DonatJR) [reply project_overview.md]: I thought about the same thing but came to the decision that in this instance I'm talking more about security attributes in general than the explicit data structure. But I can change it if you want.

Comment (DonatJR) [reply project_overview.md]: I did this on purpose because of the many "and" in this sentence 😅

File: project_overview.md (hunk 6 above)
Comment (DonatJR) [reply project_overview.md]: Again, did this to not write "and" too often. Still want me to change it?

Comment (SailReal) [reply project_overview.md]: It's okay the way it is (in both cases).

File: project_overview.md (hunk 4 above)
Comment (SailReal) [reply project_overview.md]: It's good so!
Thought it is exactly the `SECURITY_ATTRIBUTES`, because it's not it fits the way it is.

**Pull request merged**